[zoneminder] Replacement for ZoneMinder binding

mhilbush opened the pull request:
This PR is a replacement for PR #7412, which was submitted against the 2.5.x branch. The 2.5.x version of the binding has been reviewed by @cpmeister and @fwolter. There are still a couple recent changes that are awaiting review by @fwolter .

**Background**
The current OH2 ZoneMinder binding (org.openhab.binding.zoneminder) is unmaintained.
It also has a number of shortcomings:

- There are 6266 SLOC (as counted by SLOCCount). This count includes the zoneminder binding itself, as well as the zoneminder4j library that the binding uses.
- The binding was written for older versions of Zoneminder. Since then, the Zoneminder API has evolved considerably, such that much of this binding's functionality can now be accomplished much more easily with version 2.0 of the Zoneminder API.
- The current binding doesn't support Zoneminder's latest JWT-based authentication scheme, which improves security over the previous authentication methods. Anyone who has enabled authentication in recent versions of Zoneminder is unable to use the old binding.
- The current binding (and the library it uses) do not use null annotations. Adding null annotations is a significant amount of work.
- There is an abandoned 2-year old PR #3583 that added 4429 LOC and deleted 1649 LOC. That PR was closed a year ago. I took a look at the PR's review comments, as well as the code. Resurrecting it would be a monumental amount of work.

**Implications**
I wrote a new version of the binding using ZoneMinder version 1.34.0 and Zoneminder API version 2.0 as a base. My binding is 1698 SLOC (as counted by SLOCCount) versus the current binding's 6266 SLOC. It replaces the old binding in its entirety. It also is not backwards compatible with ZoneMinder versions prior to 1.34.0. Users of the old binding who upgrade to OH3 will need to upgrade their ZoneMinder to use this new binding.

The new binding has been in use for quite a while in 2.5.x. See community post here (https://community.openhab.org/t/new-zoneminder-binding-for-zoneminder-versions-1-34-0/95112).

Signed-off-by: Mark Hilbush <mark@hilbush.com>

jeremystucki commented:
Is there a way to keep the commit history? (it says that the main branch is unrelated to the 2.5.x branch)

Hilbrand commented:
You'll have a conflict on the `.classpath` because that file has been removed from GitHub.

cpmeister commented on bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java:
```diff
@@ -0,0 +1,303 @@
+
+    private void refreshImage() {
+        if (isLinked(CHANNEL_IMAGE)) {
+            scheduler.execute(this::getImage);
```
`refreshImage()` is already called asynchronously, why does `getImage()` need to be asynchronous then?

mhilbush replied:
My original thinking was that I didn't want `refreshImage` to block waiting for the image. But, I can't think of a reason why that would be necessary. I've removed it.

cpmeister commented on bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java:
```diff
@@ -0,0 +1,303 @@
+    private @Nullable ScheduledFuture<?> imageRefreshJob;
+    private @Nullable ScheduledFuture<?> alarmOffJob;
+
+    private final Map<String, State> monitorStatusCache = Collections.synchronizedMap(new HashMap<>());
```
```suggestion
    private final Map<String, State> monitorStatusCache = new ConcurrentHashMap<>();
```

mhilbush replied:
Done

cpmeister commented on bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,91 @@
+			<label>Monitor Id</label>
+			<description>Id of the monitor</description>
+		</parameter>
+		<parameter name="imageRefreshInterval" type="integer" unit="s" required="false">
+			<label>Image Refresh Interval</label>
+			<description>Interval in seconds with which monitor image is refreshed</description>
+		</parameter>
```
If it isn't required what is the default value?
Also shouldn't this have a minimum value?

mhilbush replied:
No, the default is to not refresh the image, in which case you leave the field blank.

Yes, I'll add a minimum.

cpmeister commented on bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,217 @@
+		<item-type>Number</item-type>
+		<label>Event Frames</label>
+		<description>Number of frames in the event</description>
+		<state readOnly="true" pattern="%.0f"></state>
```
I'm not sure that you can have fractional frames.
```suggestion
		<state readOnly="true" pattern="%d"></state>
```

mhilbush replied:
Correct. Fixed.

cpmeister commented on bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,217 @@
+		<item-type>Number</item-type>
+		<label>Event Alarm Frames</label>
+		<description>Number of alarm frames in the event</description>
+		<state readOnly="true" pattern="%.0f"></state>
```
same here
```suggestion
		<state readOnly="true" pattern="%d"></state>
```

mhilbush replied:
Fixed

cpmeister commented on bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,217 @@
+		<config-description-ref uri="thing-type:zoneminder:monitor"/>
+	</thing-type>
+
+	<channel-type id="id">
+		<item-type>String</item-type>
+		<label>ID</label>
+		<description>Monitor ID</description>
+		<state readOnly="true" pattern="%s"></state>
+	</channel-type>
```
Should this be a thing property instead?

mhilbush replied:
If you change the monitor id in Zoneminder, you should be able to change it here as well. Otherwise, you would need to delete and readd the thing.

fwolter replied:
How would a user do this, as it is read only? Isn't this a configuration parameter, either?

mhilbush replied:
Thank you for making me rethink this. I'll try to clarify, then we can figure out what to do.

- The above channel contains the monitor id returned by the Zoneminder API. I made it a channel because the user likely will want to include the monitor id in their sitemap or HABpanel UI. As the monitor id is returned by the Zoneminder API, and cannot be changed by the user, I believe it should be readonly.

- As for the config parameter, when creating a monitor thing manually, or using a `.things` file, the user needs to specify the monitor id as part of defining the thing. When the monitor thing is created through discovery, it is set by the binding. Also, the monitor id the representation property.

So, what to do? I'm not sure how to resolve this without impacting one of the two points above. WDYT?

fwolter replied:
Your arguments are valid. If I see correctly, the State of the Channel `id` is updated if the user changes the config parameter `monitorId`. I think it's good to go.

mhilbush replied:
Perfect. Thanks!

cpmeister reviewed (changes requested):
Looks pretty good, just have a few questions.

mhilbush commented on bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmAuth.java:
```diff
@@ -0,0 +1,172 @@
+        if (response != null) {
+            Gson gson = bridgeHandler.getGson();
+            AuthResponseDTO auth = gson.fromJson(response, AuthResponseDTO.class);
+            if (auth != null && auth.exception == null && auth.exception == null && auth.accessToken != null) {
```
No, no. I need to really make sure it's null. lol

Fixed.

mhilbush commented on bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmAuth.java:
```diff
@@ -0,0 +1,172 @@
+
+    private final Logger logger = LoggerFactory.getLogger(ZmAuth.class);
+
+    private ZmBridgeHandler bridgeHandler;
+    private String authContent;
```
Done

mhilbush commented on bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/discovery/MonitorDiscoveryService.java:
```diff
@@ -0,0 +1,128 @@
+    @Override
+    public void setThingHandler(@Nullable ThingHandler handler) {
+        if (handler instanceof ZmBridgeHandler) {
+            ((ZmBridgeHandler) handler).setDiscoveryService(this);
```
I don't know. It's done this way in a number of other bindings. And, I don't think we know the bridge handler in activate.

mhilbush commented on bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,635 @@
+    }
+
+    public @Nullable String executePost(String url, String content, String contentType) {
+        try (ByteArrayInputStream inputStream = new ByteArrayInputStream(content.getBytes())) {
```
Agreed. Done.

cpmeister commented on bundles/org.openhab.binding.zoneminder/README.md:
```diff
@@ -1,184 +1,277 @@
+| Refresh Interval               | refreshInterval             | Required  | Frequency at which monitor status will be updated. |
+| Default Alarm Duration         | defaultAlarmDuration        | Required  | Can be used to set the default alarm duration on discovered monitors. |
+| Default Image Refresh Interval | defaultImageRefreshInterval | Optional  | Can be used to set the image refresh interval on discovered monitors. Leave empty to not set an image refresh interval. |
+| Monitor Discovery Enabled      | discoveryEnabled            | Required  | Enable/disable the automatic discovery of monitors. Default is enabled. |
+| Monitor Discovery Interval     | discoveryInterval           | Required  | Frequency at which the binding will try to discover monitors. |
```
The config xml specifies that these are not required.

mhilbush replied:
Fixed

cpmeister commented on bundles/org.openhab.binding.zoneminder/README.md:
```diff
@@ -1,184 +1,277 @@
+| Port Number                    | portNumber                  | Optional  | Port number if not on ZoneMinder default port 80. |
+| Use Default Url Path           | useDefaultUrlPath           | Required  | If ON, use ZoneMinder default of /zm; if OFF, use value from urlPath parameter. |
+| Url Path                       | urlPath                     | Optional  | Use this parameter when not using the ZoneMinder default path of /zm. |
+| Refresh Interval               | refreshInterval             | Required  | Frequency at which monitor status will be updated. |
```
This one too

mhilbush replied:
Also fixed.

cpmeister reviewed (approved):
LGTM

cpmeister commented:
Looks like you need to run spotless on the config.xml

mhilbush commented:
Yeah, I was just looking at that. I didn't think I made a change that would've tweaked the formatting. I'll fix now.

mhilbush commented:
Ok, the build issue should be fixed.

But, I need to look at this.
```
[WARNING] Code Analysis Tool has found:
 0 error(s)!
 2 warning(s)
 0 info(s)
[WARNING] OH-INF.config.config.xml:[0]
Unused configuration reference with uri - thing-type:zoneminder:monitor
[WARNING] OH-INF.config.config.xml:[0]
Unused configuration reference with uri - thing-type:zoneminder:server
```

mhilbush commented:
I don't understand what's causing the above warnings. I've never seen these before. In testing the binding, it's handling the config properly.

mhilbush commented:
And now I need to fix the merge conflict with CODEOWNERS...

fwolter commented on bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java:
```diff
@@ -0,0 +1,303 @@
+        logger.debug("Monitor {}: Received command '{}' for channel '{}'", monitorId, command, channelUID.getId());
+        ZmBridgeHandler localHandler = bridgeHandler;
+        if (localHandler == null) {
+            logger.info("Monitor {}: Can't execute command because bridge handler is null", monitorId);
```
This could be debug or warn.

mhilbush replied:
Probably should be warn.

fwolter commented on bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,90 @@
+			<description>Duration in seconds after which the alarm will be turned off</description>
+			<default>60</default>
+		</parameter>
+		<parameter name="defaultImageRefreshInterval" type="integer" required="false" groupName="config-info">
```
```suggestion
		<parameter name="defaultImageRefreshInterval" type="integer" unit="s" required="false" groupName="config-info">
```

mhilbush replied:
Fixed

mhilbush commented on bundles/org.openhab.binding.zoneminder/README.md:
```diff
@@ -1,184 +1,277 @@
+| Host                           | host                        | Required  | Host name or IP address of the ZoneMinder server. |
+| Use secure connection          | useSSL                      | Required  | Use http or https for connection to ZoneMinder. Default is http. |
+| Port Number                    | portNumber                  | Optional  | Port number if not on ZoneMinder default port 80. |
+| Use Default Url Path           | useDefaultUrlPath           | Required  | If ON, use ZoneMinder default of /zm; if OFF, use value from urlPath parameter. |
```
Sorry @fwolter. I had responded to your question in the old PR, then closed the PR before you had a chance to see my response.

Here's what I said in the original PR. Perhaps you have a suggestion for how to do it differently.

> This is how I wanted to do it originally. However, if you define the parameter to have a default, I don't believe it's then possible to leave the parameter empty. If you leave it empty, OH will just insert the default. As there are some Zoneminder installations that need the path to be empty, this was the only way I could think to do it.

mhilbush replied:
So, yes, the default Zoneminder installation would be:
```
http://localhost/zm/path/to/api/endpoint
```

And an alternate would be:
```
http://localhost/path/to/api/endpoint
```

How obvious!!! Sometimes you can't see the forest for the trees. lol

I'll make the change.

mhilbush commented on bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/action/ZmActions.java:
```diff
@@ -0,0 +1,136 @@
+        }
+    }
+
+    private static IZmActions invokeMethodOf(@Nullable ThingActions actions) {
+        if (actions == null) {
+            throw new IllegalArgumentException("actions cannot be null");
+        }
+        if (actions.getClass().getName().equals(ZmActions.class.getName())) {
+            if (actions instanceof IZmActions) {
+                return (IZmActions) actions;
+            } else {
+                return (IZmActions) Proxy.newProxyInstance(IZmActions.class.getClassLoader(),
+                        new Class[] { IZmActions.class }, (Object proxy, Method method, Object[] args) -> {
+                            Method m = actions.getClass().getDeclaredMethod(method.getName(),
+                                    method.getParameterTypes());
+                            return m.invoke(actions, args);
+                        });
+            }
+        }
+        throw new IllegalArgumentException("Actions is not an instance of ZmActions");
+    }
```
Since DSL rules will still be supported in OH3, I don't recall seeing anything that fixes the root cause of the issue. And the [issue in core](https://github.com/openhab/openhab-core/issues/1265) is still open. But @cweitkamp would know for sure. Let wait for his reply.

cweitkamp replied:
Please correct me if I am wrong, I am not really up-to-date right now: IIRC there was no real confirmation for `ThingActions` are working correctly in OH3 NGRE and we wanted to keep those workarounds and bulk remove them later.

mhilbush commented on bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/action/ZmActions.java:
```diff
@@ -0,0 +1,136 @@
+        logger.debug("ZmActions: Action 'TriggerAlarm' called");
+        ZmMonitorHandler localHandler = handler;
+        if (localHandler == null) {
+            logger.info("ZmActions: Action service ThingHandler is null!");
```
Some bindings use info, but most use warn. I'll make it warn.

mhilbush commented on bundles/org.openhab.binding.zoneminder/README.md:
```diff
@@ -1,184 +1,277 @@
-String zmMonitor1_EventCause 		"Event Cause [%s]"	 		<switch>	{channel="zoneminder:monitor:ZoneMinderSample:monitor-1:event-cause"}
-String zmMonitor1_Function 		"Function [%s]" 					{channel="zoneminder:monitor:ZoneMinderSample:monitor-1:function"}
-Switch zmMonitor1_CaptureState	 	"Capture Daemon [%s]" 			<switch>	{channel="zoneminder:monitor:ZoneMinderSample:monitor-1:capture-daemon"}
-Switch zmMonitor1_AnalysisState 	"Analysis Daemon [%s]" 			<switch>	{channel="zoneminder:monitor:ZoneMinderSample:monitor-1:analysis-daemon"}
-Switch zmMonitor1_FrameState		"Frame Daemon [%s]"			<switch>	{channel="zoneminder:monitor:ZoneMinderSample:monitor-1:frame-daemon"}
+The binding requires ZoneMinder version 1.34.0 or greater, and API version 2.0 or greater.
+The API must be enabled in the ZoneMinder configuration using the **OPT_USE_API** parameter.
 
+The binding can access ZoneMinder with or without authentication.
+If you have authentication enabled in ZoneMinder, set the **user** and **pass** 
+config paremeters in the *server* thing.
```
Agreed. Merged.

mhilbush commented:
@fwolter I addressed most of your feedback in the latest commit. For the remaining ones, I've added some comments and questions. Thanks for your help!

mhilbush commented:
I've reworked the config parameters and the documentation.

fwolter reviewed (approved):
LGTM

mhilbush commented:
Thanks for your help on this @fwolter and @cpmeister!!
